daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -401,15 +378,85 @@ object ImmArray extends ImmArrayInstances {
 
   private[ImmArray] final class IACanBuildFrom[A]
       extends CanBuildFrom[ImmArray[_], A, ImmArray[A]] {
-    override def apply(from: ImmArray[_]) = newBuilder[A]
-    override def apply() = newBuilder[A]
+    override def apply(from: ImmArray[_]) = newBuilder[A]()
+    override def apply() = newBuilder[A]()
   }
 
   implicit def `ImmArray canBuildFrom`[A]: CanBuildFrom[ImmArray[_], A, ImmArray[A]] =
     new IACanBuildFrom
 
-  def newBuilder[A]: mutable.Builder[A, ImmArray[A]] =
-    mutable.ArraySeq.newBuilder[A].mapResult(ImmArray.unsafeFromArraySeq)
+  private[this] def copyArraySeq[A](
+      src: mutable.ArraySeq[A],
+      srcStart: Int,
+      dst: mutable.ArraySeq[A],
+      dstStart: Int,
+      len: Int,
+  ): Unit = {
+    var i = srcStart
+    var j = dstStart
+    val end = srcStart + len
+    while (i < end) {
+      dst(j) = src(i)
+      i += 1
+      j += 1
+    }
+  }
+
+  def newBuilder[A](initialCapacity: Int = 4): mutable.Builder[A, ImmArray[A]] =
+    new mutable.Builder[A, ImmArray[A]] {
+      private[this] var size = 0
+      private[this] var array = new mutable.ArraySeq[A](initialCapacity max 1)
+
+      override def +=(elem: A): this.type = {
+        if (size == array.length) {
+          val newArray = new mutable.ArraySeq[A](size * 2)
+          copyArraySeq(array, 0, newArray, 0, size)
+          array = newArray
+        }
+        array(size) = elem
+        size += 1
+        this
+      }
+
+      override def clear(): Unit = size = 0
+
+      override def result(): ImmArray[A] =
+        if (size == 0) {
+          empty
+        } else {
+          val result = new ImmArray[A](0, size, array)
+          array = null
+          result
+        }
+    }
+
+  def newReverseBuilder[A](initialCapacity: Int = 4): mutable.Builder[A, ImmArray[A]] =
```

Comment:
Do we have good tests for this? This seems like a prime candidate for off-by-one errors (although I believe this is indeed correct).

Reply:
I add a test.